modules/pam_mkhomedir/mkhomedir_helper.c

```diff
@@ -26,275 +26,346 @@
-         char pointed[PATH_MAX] = {};
-
-         pointedlen = readlink(newsource, pointed, sizeof(pointed) - 1);
+	    free(pointed);
 #endif
+	    goto go_out;
+	 }
 
-	 if (pointedlen >= 0) {
-            if(symlink(pointed, newdest) == 0)
-            {
-               if (lchown(newdest, pwd->pw_uid, pwd->pw_gid) != 0)
-               {
-                   pam_syslog(NULL, LOG_DEBUG,
-			      "unable to change perms on link %s: %m", newdest);
-                   closedir(d);
+	 if (fchownat(parent->fd, dent->d_name, pwd->pw_uid, pwd->pw_gid,
+		      AT_SYMLINK_NOFOLLOW) != 0)
+	 {
+	    pam_syslog(NULL, LOG_DEBUG,
+		       "unable to change perms on link %s: %m", dent->d_name);
 #ifndef PATH_MAX
-		   free(pointed);
+	    free(pointed);
 #endif
-                   free(newsource);
-                   free(newdest);
-                   return PAM_PERM_DENIED;
-               }
-            }
+	    retval = PAM_PERM_DENIED;
+	    goto go_out;
+	 }
 #ifndef PATH_MAX
-	    free(pointed);
+	 free(pointed);
 #endif
-         }
-         free(newsource);
-         free(newdest);
-         continue;
       }
+      retval = PAM_SUCCESS;
+      goto go_out;
+   }
 
-      /* If it's not a regular file, it's probably not a good idea to create
-       * the new device node, FIFO, or whatever it is. */
-      if (!S_ISREG(st.st_mode))
-      {
-         free(newsource);
-         free(newdest);
-         continue;
-      }
+   /* If it's not a regular file, it's probably not a good idea to create
+    * the new device node, FIFO, or whatever it is. */
+   if (!S_ISREG(st.st_mode))
+   {
+      retval = PAM_SUCCESS;
+      goto go_out;
+   }
 
-      /* Open the source file */
-      if ((srcfd = open(newsource, O_RDONLY)) < 0 || fstat(srcfd, &st) != 0)
-      {
+   /* Open the source file */
+   if ((srcfd = open(newsource, O_RDONLY)) < 0 || fstat(srcfd, &st) != 0)
+   {
+      pam_syslog(NULL, LOG_DEBUG,
+		 "unable to open or stat src file %s: %m", newsource);
+      retval = PAM_PERM_DENIED;
+      goto go_out;
+   }
+
+   /* Open the dest file */
+   if ((destfd = openat(parent->fd, dent->d_name,
+			O_WRONLY | O_CREAT | O_EXCL | O_NOFOLLOW, 0600)) < 0)
+   {
+      retval = errno == EEXIST ? PAM_SUCCESS : PAM_PERM_DENIED;
+      if (retval != PAM_SUCCESS)
          pam_syslog(NULL, LOG_DEBUG,
-		    "unable to open or stat src file %s: %m", newsource);
-         if (srcfd >= 0)
-            close(srcfd);
-         closedir(d);
+		    "unable to open dest file %s: %m", dent->d_name);
+      goto go_out;
+   }
 
-         free(newsource);
-         free(newdest);
+   /* Set the proper ownership and permissions for the module. We make
+      the file a+w and then mask it with the set mask. This preserves
+      execute bits */
+   if (fchmod(destfd, (st.st_mode | 0222) & (~u_mask)) != 0 ||
+       fchown(destfd, pwd->pw_uid, pwd->pw_gid) != 0)
+   {
+      pam_syslog(NULL, LOG_DEBUG,
+		 "unable to change perms on copy %s: %m", dent->d_name);
```

Comment:
Shouldn't ut be
```
"unable to change perms on copy %s: %m",
parent->path, dent->d_name);
```
?

Reply:
And also here, thanks for your careful review!